### PR TITLE
feat: add Nix development shell

### PR DESCRIPTION
**Pull request by pulsastrix**
Adds a Nix based development shell for better reproducibility (and easier setup of dependencies).

Assuming nix and direnv are activated and set up, the shell can be automatically activated by entering the directory (after a `direnv allow`)
Alternatively, you can enter the shell manually using `nix --extra-experimental-features nix-command --extra-experimental-features flakes  develop --impure .` (or just `nix develop --impure .` if the corresponding features are already enabled by your nix.conf).

### CODE REVIEW

**Review by JKRhb — COMMENTED** [review]
Thank you, @pulsastrix, I now tested it out locally on macOS and it works :) Only `dirvenv` wasn't working in the beginning but applying the fix discussed in https://github.com/direnv/direnv/issues/64 solved the problem for me. Also, I had to set `doCheck` to `false` in the case of PSync as there are a lot of errors occurring with the message “missing symbol called” – I assume you didn't get these errors on your machine, right?

Maybe you could add a few lines on the usage of Nix and direnv to the README? And would it make sense to also use the Nix setup for the CI?

**Review by tanim-ics — APPROVED** [review]

**Comment by pulsastrix** [comment]
>  Also, I had to set `doCheck` to `false` in the case of PSync as there are a lot of errors occurring with the message “missing symbol called” – I assume you didn't get these errors on your machine, right?
> Edit: Full error log for PSync

Correct, it did work without issue on my machines - probably a macOS exclusive problem?
I'll recheck tomorrow.

> Maybe you could add a few lines on the usage of Nix and direnv to the README? 

Will do.

> And would it make sense to also use the Nix setup for the CI?

_Probably_, although I am unsure how large the impact on build times would be, I'll have a look into it.
There are [solutions](https://github.com/DeterminateSystems/magic-nix-cache) that allow using the GitHub Actions cache, which might reduce the number of times that dependencies have to be rebuilt in order to set up the environment.

One additional thing - we might also be able to quickly set up required background services using the [Process feature of devenv.sh](https://devenv.sh/processes/), one such example might be `nfd`. 
It would probably also make sense to look into that before merging, I hope that I'll find the time to do so tomorrow.

**Comment by pulsastrix** [comment]
The installation instructions have now been expanded with a small guide on how to compile using the Nix environment.

Additionally, I have started work on a Nix development guide that covers the most important parts of using it for development and changing the Nix environment if necessary.

Feel free to make suggestions for topics that should also be covered in the developer guide and aren't yet present.

File: Install.md [hunk 1]
```diff
@@ -137,6 +144,70 @@ otool -L /usr/local/lib/libPSync.dylib
...
+IceFlow developers might also want to refer to our developer documentation regarding the nix configuration, which can
+be found in [Nix-Developer-Guide.md](doc/Nix-Developer-Guide.md).
+
+For the purposes of IceFlow development, the IceFlow Nix environment can be considered as a rough equivalent to
+virtualenvs in Python.
+
+### Requirements
+
+In order to compile and run IceFlow using Nix, the following system requirements must be fulfilled:
+1. A Linux or MacOS based system (Windows might work through the WSL, but is untested, refer to
```
Comment (JKRhb) [inline Install.md:164]:
```suggestion
1. A Linux or macOS based system (Windows might work through the WSL, but is untested, refer to
```

File: doc/Nix-Developer-Guide.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+without an absolute path:
+![Nix toolchain configuration in CLion](nix_clion.png)
+
+
+## Changing the Nix environment
+
+### Updating dependencies
+
+For updates from the Nix package repositories:
+`nix flake update` (will update flake.lock)
```
Comment (JKRhb) [inline doc/Nix-Developer-Guide.md:49]:
```suggestion
`nix flake update` (will update `flake.lock`)
```

File: flake.nix [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+            sha256 = "sha256-IY3hq06l4MYpNw2GKY9g9nLyY/yvuNKl10BYz/CJJyg=";
+          };
+
+          nativeBuildInputs = with prev; [ pkg-config wafHook python3 ];
+          buildInputs = [ ndn-cxx prev.sphinx prev.openssl ];
+
+          wafConfigureFlags = [
+            "--boost-includes=${prev.boost179.dev}/include"
+            "--boost-libs=${prev.boost179.out}/lib"
+            #"--without-tests"
```
Comment (JKRhb) [inline flake.nix:68]: Could this line be removed? Or should we keep it?

```suggestion
            #"--without-tests"
```

File: flake.nix [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+            owner = "named-data";
+            repo = "ndn-cxx";
+            rev = "18ccbb3b1f600d913dd42dd5c462afdac77e37e0";
+            hash = "sha256-yHsp6dBq2kMsubJrn77qeQ9Ah+Udy7nE9eWBX2smemA="; 
+            fetchSubmodules = true; 
+          };
+
+        });
+
+        # Update nfd to newest commit.
```
Comment (JKRhb) [inline flake.nix:32]:
```suggestion
        # Update nfd to specific commit.
```

**Review by JKRhb — APPROVED** [review]
LGTM already :) See some minor comments/suggestions below.

File: flake.nix (hunk 3 above)
Comment (pulsastrix) [reply flake.nix:68]: I'd keep it, we should probably re-enable tests as soon as they no longer fail on macOS, and keeping the flag in makes it easier to re-enable.

**Pull request merged**